flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -390,6 +453,23 @@ private boolean isPartialResourceConfigured(JobGraph jobGraph) {
 			rpcService.getExecutor());
 	}
 
+	private CompletableFuture<JobManagerRunner> createJobManagerRunner(JobGraph jobGraph, StreamManagerAddress address) {
+		final RpcService rpcService = getRpcService();
+
+		return CompletableFuture.supplyAsync(
+			CheckedSupplier.unchecked(() ->
+				new StreamingJobManagerRunnerFactory(address).createJobManagerRunner(
```

Comment:
Why do we need to make sm address as a parameter of factory, rather than parameter of createJobManagerRunner() method.

Reply:
If we add a parameter in method createJobManagerRunner(), we also need to change the interface:
```
/**
 * Factory for a {@link JobManagerRunner}.
 */
@FunctionalInterface
public interface JobManagerRunnerFactory {

	JobManagerRunner createJobManagerRunner(
		JobGraph jobGraph,
		Configuration configuration,
		RpcService rpcService,
		HighAvailabilityServices highAvailabilityServices,
		HeartbeatServices heartbeatServices,
		JobManagerSharedServices jobManagerServices,
		JobManagerJobMetricGroupFactory jobManagerJobMetricGroupFactory,
		FatalErrorHandler fatalErrorHandler) throws Exception;
}
```
Thus, the other implementation of this interface will also need to be changed.
In my application, I implement another factory that accepts the sm address in its constructor. In this way, there is nothing to be changed in `JobManagerRunnerFactory` and `JobManagerRunnerImpl`.